feat: Add Next.js plugin

kripod opened the pull request:
Resolves #98 and helps with https://github.com/kripod/glaze/issues/6#issuecomment-604541326.

mattsjones commented on packages/next-plugin-treat/index.js:
```diff
@@ -0,0 +1,19 @@
+
+module.exports = (pluginOptions = {}) => (nextConfig = {}) => {
+  return Object.assign({}, nextConfig, {
+    webpack(config, { isServer }) {
```
Where does `webpack` come from here?

kripod replied:
It's an ES2015 [method definition](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Method_definitions) shorthand. The implementation was heavily inspired by the official [`@next/mdx`](https://github.com/zeit/next.js/tree/canary/packages/next-mdx) plugin.

mattsjones replied:
Right... I completely read that wrong. Dunno if it's me or maybe it's too clever 😅

kripod commented:
Unfortunately, I'm not skilled at testing, but I'll try my best, giving feedback about the progress.

mattsjones commented:
Have a look here for the gatsby test implementation 
https://github.com/seek-oss/treat/blob/master/packages/gatsby-plugin-treat/tests/styling.test.js

maraisr commented on packages/next-plugin-treat/index.js:
```diff
@@ -0,0 +1,19 @@
+            outputCSS: !isServer,
+          }),
+        ),
+        new MiniCssExtractPlugin(),
```
This isnt needed - nextjs uses minicss itself, so the plugin is already configured.

https://github.com/zeit/next.js/blob/ea7c1e1dcf5b4019365e1ab3ff989801c958f0e9/packages/next/build/webpack/config/blocks/css/index.ts#L302-L317

maraisr commented on packages/next-plugin-treat/index.js:
```diff
@@ -0,0 +1,19 @@
+const MiniCssExtractPlugin = require('mini-css-extract-plugin');
+const TreatPlugin = require('treat/webpack-plugin');
+
+module.exports = (pluginOptions = {}) => (nextConfig = {}) => {
```
You may also want to call "maybe" passed in `nextConfig.webpack`, so chaining still works.

https://github.com/zeit/next.js/blob/canary/packages/next-mdx/index.js#L17-L19

maraisr commented on packages/next-plugin-treat/package.json:
```diff
@@ -0,0 +1,22 @@
+  },
+  "license": "MIT",
+  "dependencies": {
+    "mini-css-extract-plugin": "^0.9.0"
```
Please mark this as a peer

kripod replied:
Thank you for all the valuable feedback, I've fixed the 3 issues.

markdalgleish replied:
I'm not sure this is right. Consumers aren't installing this package themselves—it's actually a dependency of Next—so they'll get peer dependency warnings, right? I'm concerned that this isn't the right way for a Next plugin to output CSS. Tried finding other Next plugins that import `mini-css-extract-plugin` directly and can't find any examples.

kripod replied:
I’m not quite familiar with the architecture of Next. About the peer dependencies, I had the same concerns and could make it opt-in with a `peerDependenciesMeta` field. Alternative methods for encapsulating the `mini-css-extract-plugin` would be great.

There’s an RFC discussing the form of upcoming Next plugins at https://github.com/zeit/next.js/issues/9133. Unfortunately, it seems to lack webpack config modification support intentionally, due to concerns about performance.

Plugins are not widely used within the Next ecosystem. Current solutions aren’t highly documented, which explains the low amount of plugins available.

kripod commented:
Firstly, I've added an example package, copying the `src` folder directly from the `gatsby-plugin-treat-example` package. Static CSS extraction issues are now fixed.

kripod commented:
@mattsjones I've adopted the integration tests for the Next.js plugin. Unfortunately, I couldn't make a snapshot on Windows. I would be grateful if someone else could generate the snapshot on a UNIX-based system by using `yarn test`.

kripod commented:
@mattsjones I've adopted the integration tests for the Next.js plugin. Please review the changes made!

mattsjones commented on packages/next-plugin-treat/index.js:
```diff
@@ -0,0 +1,21 @@
+    webpack(config, options) {
+      config.plugins.push(
+        new TreatPlugin(
+          Object.assign({}, pluginOptions, {
```
How does mini-extract get applied?

kripod replied:
It turns out that Next.js applies it without intervention. Not sure about how, but both debug and production builds work properly. I’ve tried to add mini-extract explicitly, but it always caused trouble and resulted in bad renders.

mattsjones replied:
You need to set `outputLoaders` though. Otehrwise treat will default to using `style-loader`.

maraisr replied:
You need outputLoaders @kripod - think it only worked because of style-loader.

What you had before, was perfect.

kripod replied:
Static CSS is now extracted, but only in production, [as Next.js does it internally](https://github.com/zeit/next.js/blob/a81d7620a40b2006cb6726236402fc3f82f65ab8/packages/next/build/webpack/config/blocks/css/index.ts#L298-L320).

kripod commented:
Fixes have been made again, I would like to request a review!

aarongarciah commented on packages/next-plugin-treat-example/src/components/Button.js:
```diff
@@ -0,0 +1,4 @@
+import React from 'react';
+import {button} from './Button.treat'
+
+export const Button = ({...otherProps}) => <button {...otherProps} className={button}/>;
```
```suggestion
export const Button = (props) => <button {...props} className={button}/>;
```

kripod replied:
The example was adopted straight from the Gatsby counterpart.

kripod commented:
After some investigation, I've observed that Next.js [uses the `getClientStyleLoader` function internally](https://github.com/zeit/next.js/blob/f9cd7c581f5f952e2ab6064b9e750613f09cd948/packages/next/build/webpack/config/blocks/css/loaders/global.ts#L13-L22) for loading styles with webpack.

I tested the following code with glaze and Next seems to pick up the styles properly, even during SSR:

https://github.com/kripod/glaze/blob/ad72f13dffc31af28309012ce971e4f66625b629/packages/example-nextjs/next.config.js

kripod commented:
May I ask whether I could help with anything to advance the state of this PR? 😊

maraisr commented:
hi @kripod - I think from my perspective this is looking great!

@markdalgleish and co are very slow to respond to OSS issues and pull requests - they tend to "balance the priorities".

I think your best bet is to ping @markdalgleish directly on twitter?

kripod commented:
Thank you for the review, @maraisr!

I'm not trying to rush anything, especially in these hard times. Take your time to merge PRs as soon as you feel comfortable.

markdalgleish commented:
> For TypeScript users, this requires `@types/webpack` to be bundled as a dependency.

Sorry, can you explain this one to me? Why does TypeScript care about this file?

kripod commented:
> 
> 
> > For TypeScript users, this requires `@types/webpack` to be bundled as a dependency.
> 
> Sorry, can you explain this one to me? Why does TypeScript care about this file?

Honestly, I'm not sure why it would be required anymore. I removed it from the dependencies, eagerly waiting for the new CI build to pass.

markdalgleish commented:
Heads up—for consistency with the Next.js ecosystem, I've renamed this from `next-plugin-treat` to `next-treat`.

kripod commented:
Good idea, especially with [their RFC for plugins](https://github.com/zeit/next.js/issues/9133) with a new purpose and format.

markdalgleish commented:
Hmmm, actually, that link you sent seems to indicate they're wanting to move to a `next-plugin-` prefix. I should probably change it back.

markdalgleish commented:
Or is the point that the `next-plugin-` prefix will change the behaviour of the plugin from Next's perspective, so we should avoid it for now?

kripod commented:
Yes, exactly.

The plugin may not be compatible with the upcoming Next.js plugin architecture. Currently, they are going for a solution which doesn’t integrate webpack config modification.

markdalgleish commented:
@kripod Looking good. Happy for this to go out?

kripod commented:
Absolutely! Thanks for everyone taking the time to work all this out.
